### PR TITLE
This code use to calculate the middle element of the linked list

### PR DESCRIPTION
**Pull request by Vikash-8090-Yadav**
#### Description of Change

issue No:  #1866 

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1867"></a>

### CODE REVIEW

File: data_structures/MIddleElement_S.ll.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
+/**
+ * @file MiddleEle_SingLL.cpp
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:2]:
```suggestion
 * @file
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    n2->next = n3;
+    n3->next=n4;
+    n4->next = n5;
+    
+    int len1 = length(head);
+    midpoint(len1,head);
+    return 0;
+
+
+}
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:119]:
```suggestion
}

```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Hey there! Thank you for submitting a PR! 😄 Sadly, the documentation is not up to our [**standards**](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md). Please update the file accordingly. **Let us know if you need any help here or in our [Discord server](https://discord.gg/c7MnfGFGa6).** 😃

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Could you please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/4764269775?check_suite_focus=true) warnings? Let us know if you need any help here or in our [**Discord server**](https://discord.gg/c7MnfGFGa6)! 😃

**Comment by Vikash-8090-Yadav** [comment]
idk why this clang tidy is showing error btw in my ide (sublime text ) the codes work's completely fine ..
Correct if am missing something sir ..
thank u 😅

**Comment by Panquesito7** [comment]
> idk why this clang tidy is showing error btw in my ide (sublime text ) the codes work's completely fine .. Correct if am missing something sir .. thank u 😅

Yeah, it works fine, but there are some errors that should be fixed.

image

File: data_structures/MIddleElement_S.ll.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+ * slow and fast we point slow = head , and fast to the head->next
+ * and if length of the L.L is even then traverse till fast!=NULL
+ * and we move slow by one step and fast by one step and when fast
+ * became null then we just print the slow->data , (our mid element).
+ * For the odd length all the process will be same but only one diff. 
+ * isthat we need to check instead of fast!=null we have to chek 
+ * fast->next!=null and similarlly we print the slow->data
+ */
+
+ #include <iostream>
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:20]:
```suggestion
#include <iostream>   /// for IO operations
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+/**
+ * A Node class containing a value and pointer to another link
+ */
+class node{
+public:
+    int data;
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:27]: Please add a one-line description of what the variable is for (**see the example below**).

```c++
int64_t data = 0;       ///< The value/key of the node.
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    */
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+
+
+
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:93]:
```suggestion
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    n1->next = n2;
+    n2->next = n3;
+    n3->next= n4;
+    n4->next = n5;
+    
+    int len1 = length(head);
+    midpoint(len1,head);
+    return 0;
+
+
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:118]:
```suggestion
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+ * slow and fast we point slow = head , and fast to the head->next
+ * and if length of the L.L is even then traverse till fast!=NULL
+ * and we move slow by one step and fast by one step and when fast
+ * became null then we just print the slow->data , (our mid element).
+ * For the odd length all the process will be same but only one diff. 
+ * isthat we need to check instead of fast!=null we have to chek 
+ * fast->next!=null and similarlly we print the slow->data
+ */
+
+ #include <iostream> // for I/O operations
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:20]:
```suggestion
 #include <iostream>  /// for I/O operations
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * A Node class containing a value and pointer to another link
+ */
+class node{
+public:
+    int data;            // The value / key of the node . 
+    node*next;          //  pointer pointing to the next node adress.
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:28]:
```suggestion
    int data;            ///< The value / key of the node . 
    node*next;           ///<  pointer pointing to the next node adress.
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        this->data=data;
+        next=nullptr;
+    }
+};
+/*
+ * A midpoint function use to calculate the mid element of the l.l
+ */
+
+
+void midpoint(int len,node*head){
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:39]: Missing parameters documentation.

File: data_structures/MIddleElement_S.ll.cpp [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        }
+        std::cout<<slow->data<<" ";
+    }
+}
+
+/**
+ * A print function takes the head node and traverse the l.l
+ * and print the data
+ */
+void print(node*head){
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:62]: Missing parameters documentation.

File: data_structures/MIddleElement_S.ll.cpp [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    while(temp!=nullptr){
+        std::cout<<temp->data<<" ";
+        temp = temp->next;
+    }
+}
+/**
+ * A length function takes head pointer as a arg and 
+ * and len variable increment by 1 whenever we move head by 1
+ */
+int length(node*head){
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:73]: Missing parameters documentation.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
For how to document your code properly, you can see the [**typical structure of a program**](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) as a reference. That'll help you make your code well-documented and understandable. You can also read the whole contributing guidelines; it'll help as well. 😃

**Comment by Vikash-8090-Yadav** [comment]
remainder !!

File: data_structures/MIddleElement_S.ll.cpp [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+ * (two pointer approach),basically in this we make two pointer
+ * slow and fast we point slow = head , and fast to the head->next
+ * and if length of the L.L is even then traverse till fast!=NULL
+ * and we move slow by one step and fast by one step and when fast
+ * became null then we just print the slow->data , (our mid element).
+ * For the odd length all the process will be same but only one diff. 
+ * isthat we need to check instead of fast!=null we have to chek 
+ * fast->next!=null and similarlly we print the slow->data
+ */
+ #include <iostream> // for I/O operations
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:19]:
```suggestion
#include <iostream>  /// for I/O operations
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+/**
+ * A Node class containing a value and pointer to another link
+ * Containing the explicit constructor which contains
+ * @param data that always initialize to the node data every time
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:24]: This doesn't count as a parameter as classes can't have parameters. You can document the variables inside the classes, though. 🙂

File: data_structures/MIddleElement_S.ll.cpp [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+ * and we move slow by one step and fast by one step and when fast
+ * became null then we just print the slow->data , (our mid element).
+ * For the odd length all the process will be same but only one diff. 
+ * isthat we need to check instead of fast!=null we have to chek 
+ * fast->next!=null and similarlly we print the slow->data
+ */
+ #include <iostream> // for I/O operations
+
+/**
+ * A Node class containing a value and pointer to another link
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:22]:
```suggestion
 * @brief A Node class containing a value and pointer to another link
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+ * Containing the explicit constructor which contains
+ * @param data that always initialize to the node data every time 
+ * after creating node and next to nullptr. 
+ */
+class node{
+public:
+    int data;            ///< The value / key of the node
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:29]:
```suggestion
    int data = 0;            ///< The value / key of the node 
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+        this->data=data;
+        next=nullptr;
+    }
+};
+
+/**
+ * function use to calculate the mid element of the linked list.
+ * @param len is the length of the linked list .
+ * @param head stored the first node address of the linkedlist .
+ * @returns 'void'
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:41]:
```suggestion
 * @returns void
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    int data;            ///< The value / key of the node 
+    node*next;          ///<  pointer pointing to the next node adress.
+    explicit node(int data){
+        this->data=data;
+        next=nullptr;
+    }
+};
+
+/**
+ * function use to calculate the mid element of the linked list.
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:38]:
```suggestion
 * @brief Function use to calculate the mid element of the linked list
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+ * @param len is the length of the linked list .
+ * @param head stored the first node address of the linkedlist .
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:40]:
```suggestion
 * @param len length of the linked list
 * @param head stores the first node address of the LinkedList
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+        while(fast->next!=nullptr){
+            slow=slow->next;
+            fast=fast->next->next;
+        }
+        std::cout<<slow->data<<" ";
+    }
+}
+
+/**
+ * function takes the head node and traverse the l.l
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:63]:
```suggestion
 * @brief function takes the head node and traverses the l.l
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+ * and print the data .
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:64]:
```suggestion
 * and prints the data
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+            fast=fast->next->next;
+        }
+        std::cout<<slow->data<<" ";
+    }
+}
+
+/**
+ * function takes the head node and traverse the l.l
+ * and print the data . 
+ * @param head stored the first node address of the linkedlist .
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:65]:
```suggestion
 * @param head stores the first node address of the LinkedList
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+ * @returns 'void'
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:66]:
```suggestion
 * @returns void
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+    node*temp = head;
+    while(temp!=nullptr){
+        std::cout<<temp->data<<" ";
+        temp = temp->next;
+    }
+}
+
+/**
+ * function used to calculate the length of the linked
+ * list .
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:78]:
```suggestion
 * @brief Utility function to calculate the length of the linked list
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+ * @param head stored the first node address of the linkedlist .
+ * @returns len that is the length of the linked list.
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:80]:
```suggestion
 * @param head stores the first node address of the LinkedList
 * @returns the length of the linked list
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍

**Comment by Vikash-8090-Yadav** [comment]
Remainder.

File: data_structures/MIddleElement_S.ll.cpp [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+        temp=temp->next;
+        len++;
+    }
+    std::cout<<"The middle element is:";
+    return len;
+    /*
+    * returning the final length of the linked list 
+    */
+}
+
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:96]: Please add self-test cases so we can verify your program works properly (**see the [typical structure of a program](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) as a reference**). Let us know if you need any help here or in our [**Discord server**](https://discord.gg/c7MnfGFGa6). 😃

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: data_structures/MIddleElement_S.ll.cpp [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+ * slow and fast we point slow = head , and fast to the head->next
+ * and if length of the L.L is even then traverse till fast!=NULL
+ * and we move slow by one step and fast by one step and when fast
+ * became null then we just print the slow->data , (our mid element).
+ * For the odd length all the process will be same but only one diff. 
+ * isthat we need to check instead of fast!=null we have to chek 
+ * fast->next!=null and similarlly we print the slow->data
+ */
+#include <cassert>   ///  for assert
+#include <iostream> /// for I/O operations
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:20]:
```suggestion
#include <cassert>     /// for assert
#include <iostream>    /// for I/O operations
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+    n1->next = n2;
+    n2->next = n3;
+    n3->next= n4;
+    n4->next = n5; 
+    int len1 = length(head);
+    // Length of the linked list
+    assert(len1 == 5);
+    //Midpoint of linked list 
+    assert(midpoint(len1,head) == 3);
+    std::cout<<"All tests have sucessfully passed!"<<std::endl;
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:127]:
```suggestion
    std::cout << "All tests have sucessfully passed!" << std::endl;
```

File: data_structures/MIddleElement_S.ll.cpp [hunk 28]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+/**
+ * Main function :
+ * @returns 0 on exit
+ */
+int main(){
+	test();  //run self-test implementations
```
Comment (Panquesito7) [inline data_structures/MIddleElement_S.ll.cpp:135]:
```suggestion
    test();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

**Pull request closed without merging**